src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
+                }
             }
 
-            if (input == null)
+            // Coerces a scalar value
+            static object ParseValueScalar(ScalarGraphType scalarGraphType, string variableName, object value)
             {
-                return;
-            }
+                if (value == null)
+                    return null;
 
-            if (type is ScalarGraphType scalar)
-            {
-                // verify value can be converted successfully
+                object ret;
 
-                if (input is IValue value)
+                try
                 {
-                    bool conversionFailed;
-
-                    try
-                    {
-                        conversionFailed = scalar.ParseLiteral(value) == null;
-                    }
-                    catch (Exception ex)
-                    {
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{value.Value}' to '{type.Name}'", ex);
-                    }
-
-                    if (conversionFailed)
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{value.Value}' to '{type.Name}'");
+                    ret = scalarGraphType.ParseValue(value);
                 }
-                else
+                catch (Exception ex)
                 {
-                    bool conversionFailed;
-
-                    try
-                    {
-                        conversionFailed = scalar.ParseValue(input) == null;
-                    }
-                    catch (Exception ex)
-                    {
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'", ex);
-                    }
-
-                    if (conversionFailed)
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'");
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'", ex);
                 }
 
-                return;
+                if (ret == null)
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'");
+
+                return ret;
             }
 
-            if (type is ListGraphType listType)
+            static object ParseValueList(ListGraphType listGraphType, string variableName, object value)
             {
-                var listItemType = listType.ResolvedType;
+                if (value == null)
+                    return null;
 
-                if (input is IEnumerable list && !(input is string))
+                // note: a list can have a single child element which will automatically be intrepreted as a list of 1 elements. (see below rule)
+                // so, to prevent a string as being intrepreted as a list of chars (which get converted to strings), we ignore considering a string as an IEnumerable
```

Comment:
```suggestion
                // so, to prevent a string as being interpreted as a list of chars (which get converted to strings), we ignore considering a string as an IEnumerable
```

Thanks for meaningful comments.